### PR TITLE
[cherrypick][6.7]CDAP-18616 : handling initialization before provision/ deprovision

### PR DESCRIPTION
**Pull request by sahusanket**
original pr : https://github.com/cdapio/cdap/pull/14219 , https://github.com/cdapio/cdap/pull/14292

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/ProvisioningService.java [hunk 1]
```diff
@@ -131,6 +132,8 @@ public class ProvisioningService extends AbstractIdleService {
   private KeyedExecutor<ProvisioningTaskKey> taskExecutor;
   private ExecutorService contextExecutor;
 
+  private final CountDownLatch initializeLatch = new CountDownLatch(1);
```
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/ProvisioningService.java:136]: Add comment on why this latch is needed.

**Review by masoud-io — APPROVED** [review]

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/ProvisioningService.java (hunk 1 above)
Comment (sahusanket) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/ProvisioningService.java:136]: The provisioning service class is used at different places

1. Some places just use without having to use provision/deprovision
2. Some places just have a hard dependency on the Service's successful startup as they use provision/deprovision which depends on `initializeProvisionersAndExecutors` method [ Example : `programNotificationSubscriberService`]
So, for the 2nd point , to ensure `initializeProvisionersAndExecutors` is completed before they use provision/deprovision , we are using the CountDownLatch as a flag to check if the initialization is complete.

So if provision/deprovision is called before initialization, they will wait until the CountDownLatch is complete.

Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/ProvisioningService.java:136]: I think he meant a comment in the code

Comment (sahusanket) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/ProvisioningService.java:136]: My bad. added in code.

**Pull request merged**